### PR TITLE
Add patch style for oui fixed components

### PR DESCRIPTION
**Pull request by raintygao**
### Description
There are some UI components from library whose position is fixed and are not compatible with each other and also not compatible with sidecar container. Meanwhile, there is currently no way to provide config for these components at runtime.
This PR adds a hook to patch a style for all these already known components to make them compatible with sidecar.
Will consider to support config provider in OUI and refer this PR later.

### ScreenShot

image
image
image

Before: the left or right property of flyout is always to be 0. The content width of bottomBar is  always be 100%.

image
image
image

This still needs a PR in OSD to update sidecar z-index. https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6964
After: the left or right property of flyout will be updated accordingly. The bottomBar will be added paddingRight or paddingLeft style accordingly.

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test.
- [ ] New functionality has user manual doc added.
- [ ] Commits are signed per the DCO using --signoff.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/hooks/use_patch_fixed_style.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    style.appendChild(text);
+
+    textRef.current = text;
+
+    const subscription = sidecarConfig$.subscribe((config) => {
+      if (!config) return;
+      updateHeadStyle(config, textRef.current);
+    });
+    return () => {
+      subscription.unsubscribe();
```
Comment (wanglam) [inline public/hooks/use_patch_fixed_style.ts:33]: Shall we remove appended style after unmount?

**Comment by ruanyl** [comment]
@raintygao Can you also add a screenshot in the description to better show what is the problem?

**Comment by raintygao** [comment]
> @raintygao Can you also add a screenshot in the description to better show what is the problem?

Sure. Updated.

File: public/chat_header_button.tsx [hunk 2]
```diff
@@ -52,6 +53,7 @@ export const HeaderChatButton = (props: HeaderChatButtonProps) => {
   const core = useCore();
   const flyoutFullScreen = sidecarDockedMode === SIDECAR_DOCKED_MODE.TAKEOVER;
   const flyoutMountPoint = useRef(null);
+  usePatchFixedStyle();
```
Comment (SuZhou-Joe) [inline public/chat_header_button.tsx:56]: I am thinking, the patched style should only apply when the chatbot is visible right?

**Comment by raintygao** [comment]
A PR in OSD that is related.cc @ruanyl @SuZhou-Joe https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6964

File: public/chat_header_button.tsx (hunk 2 above)
Comment (raintygao) [reply public/chat_header_button.tsx:56]: Yes.

Comment (raintygao) [reply public/chat_header_button.tsx:56]: Patchstyle contains the logic that when chatbot is closed or docked takeover. When in these cases, we will update a empty style to remove patch style.

Comment (SuZhou-Joe) [reply public/chat_header_button.tsx:56]: I see, so the `sidecarConfig$` will emit a null value once the Chatbot is not visible?

**Review by wanglam — APPROVED** [review]

File: public/chat_header_button.tsx (hunk 2 above)
Comment (raintygao) [reply public/chat_header_button.tsx:56]: Sidecar config will emit a new value with isHidden property is true.Then we will update inserted style with ""

Comment (SuZhou-Joe) [reply public/chat_header_button.tsx:56]: I did not find the code that we will update the inserted style when config.hidden is true.

File: public/hooks/use_patch_fixed_style.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          padding-left: ${paddingSize}px
+      }
+        `;
+    }
+  }
+
+  // If sidecar closes or docked direction change to takeover mode, we will keep css empty to remove patch style and update.
+  requestAnimationFrame(() => {
+    text.textContent = css;
+  });
```
Comment (SuZhou-Joe) [inline public/hooks/use_patch_fixed_style.ts:87]:
```suggestion
  requestAnimationFrame(() => {
    if (config?.isHidden) { text.textContent = ''; return ; }
    text.textContent = css;
  });
```

Nit: I'd suggest explicitly making the textContent as empty.

**Review by SuZhou-Joe — APPROVED** [review]

File: public/hooks/use_patch_fixed_style.ts (hunk 3 above)
Comment (raintygao) [reply public/hooks/use_patch_fixed_style.ts:87]: I think here takeover dockedMode also takes the same effect as isHidden, should we also explicitly append?

**Pull request merged**